### PR TITLE
Jianjun and RJ - Performance Optimization

### PR DESCRIPTION
**Pull request by juneluo**
### Description
Complete the priority medium task of LEADERBOARD COMPONENT
Improve loading speed of other people’s Time Log: When I click the time bar in the leaderboard for someone, the loading speed to see their time log is really slow. I’d like to speed that up if possible. 
Remaining major open problems proposed by RJ are described in the analysis [here](https://www.dropbox.com/sh/e43le4bts99dxuw/AADHoHLWQVaf1FDHdqRgU2gNa?dl=0) 
To sum up, major points to check:
1. avoid unnecessary re-render
2. check the dependency of the hooks
3. convert the Timelog component from class to function

### Mainly changes explained:
1. set the value of prop `isDashboard` as true when the timelog component is called at the main page, remove a duplicated condition.
2. Introduce the state `summaryBarData` and redesign the passing process of the data for the SummaryBar component so that it won’t use the ones from the leader board.
3. Convert the Timelog component from class to function and use the hooks and states to control the re-render.

### How to test:
To test if the component runs normally:
Log in --> see dashboard --> click timelog page ==> shows the correct output
log out and log in as another account --> see dashboard ===>shows the correct output
add time entry ===>shows the correct output
To test the performance:
add console.log in the code and check console output to see how many times each part/function runs when open the page.

### CODE REVIEW

**Review by PrabhjotSembhi — COMMENTED** [review]
It seems faster. but I am not sure how this will test the performance 
How to test:
Log in --> see dashboard --> click timelog page --> log out and log in as another account --> see dashboard --> add time entry

**Review by ypan710 — COMMENTED** [review]
I'm not sure why for some of my dummy accounts, when I cannot select "Project/Task" under "Add Tangible Time Entry".

Screen Shot 2023-04-10 at 5 22 58 PM

**Comment by LucileTech** [comment]
> > I'm not sure why for some of my dummy accounts, when I cannot select "Project/Task" under "Add Tangible Time Entry".
> > Screen Shot 2023-04-10 at 5 22 58 PM
> 
> I was going to say, check to be sure you have Projects assigned to those accounts. The Orientation/Setup project is assigned by default though... so that should be an option there. This may be a new bug. Anyone else experiencing this?

I created a brand new account to test the @ypan710 concern but I do not have the issue.  The brand new account has the  Orientation/Setup project assigned by default.
I tested it with admin, volunteer, and new volunteer.

Screenshot 2023-04-11 at 18 47 41

**Review by Elisaudo — DISMISSED** [review]
juneluo I tested PR#762 and saw that the loading speed is better, like shown in vídeo before and after.
Before:
https://user-images.githubusercontent.com/64317219/231568163-42ffaed7-1427-41d5-90f8-9cb8cde79c95.mp4

After:
https://user-images.githubusercontent.com/64317219/231568236-8ad20648-68be-4da0-8dc4-d7ebbdea980b.mp4

**Comment by AnishPandita** [comment]
Hi @juneluo I tested your PR and noticed a significant difference in loading time. I does seem you were able to optimise it. 

https://user-images.githubusercontent.com/111321945/231629460-f153eb79-fe16-4454-8b31-8df19bdd8143.mov

**Comment by AuroraLovesWhales** [comment]
Hi! I have tested your PR#762 and the loading speed does seem to improve. Great job! 

https://user-images.githubusercontent.com/44281916/232129666-5167a5a6-3d1d-46c9-89dd-e47032d3149d.mov

**Comment by AriaYu927** [comment]
Hi @juneluo , I've tested the PR and can see the loading speed does improved, thank you.

Before:
https://user-images.githubusercontent.com/94303659/232251374-8be8c7dc-e410-4f96-89e8-2548422b0010.mp4

After:
https://user-images.githubusercontent.com/94303659/232251376-111e86d1-8bd1-41b9-9830-68360ba202db.mp4

**Comment by RJBornToWin** [comment]
Drastic improvement in performance. Commendable effort from Jianjun.

High critical issue: Clicking on the status button on the Timelog page -> Crashes the backend server. Have to restart the backend to use the front end application. I used "development" branch as back end.

Things that needs to be handled in future PRs since this is time sensitive:
1. Dashboard loading(~ 5 - 6 secs) is slower compared to Timelog (~ 2 secs).  Any possibility to adapt timelog page changes to dashboard page can be explored and adapted.
2. Timelog page cannot be reloaded for self. Data gets changed only when user changes. useEffect(callback, []) is causing this in Timelog. At least a refresh button has to be introduced within the page.
3. Comments to be added for the functions written in Timelog as it helps in maintaining the code quality on long run.
4. (Optional): There is a SummaryBar component that is directly used in Dashboard and one within Timelog. So there are 2 SummaryBar components even though one isn't visible to the user. This can be cleaned up.

**Review by vitorsaa2k — CHANGES_REQUESTED** [review]
While testing I've noticed that the feature from PR #686 has been removed / is not working, I've only tested using an Owner account. Despite that everything seems to be working fine.

**Comment by one-community** [comment]
Good catch Vitor! A lot of stuff like that has been happening with PRs lately, functions being lost and/or bugs returning that have already been fixed.

**Comment by one-community** [comment]
Sorry for my slow reply here, just now seeing this question for me. The time log tab should be default if a Volunteer or Core Team AND/OR if no task. If a person is a Manager, Mentor, Admin or Owner then the Tasks tab should be default. If ANY class type has a task, then the tasks tab should show.

**Comment by RJBornToWin** [comment]
image

@juneluo : Adding additional details. If you click on that green button/icon (status icon) on the timelog page, the app crashes.

File: src/components/LeaderBoard/Leaderboard.jsx
Comment (Vishvesh-Sheoran) [inline src/components/LeaderBoard/Leaderboard.jsx:1]: 1. In the _useEffect_ hook in line 28, the _setLeaderData_ function is not called within a try-catch block like _useDeepEffect_ in line 21. Wrap it inside a try-catch block.

2. In line 39, the _isOpen_ state variable should be declared with an initial value of _false_. This is because a value of _undefined_ can cause unexpected behavior.

3. In the _Modal_ component in line 64, add an _aria-labelledby_ attribute to provide a label for the modal dialog. This improves accessibility by associating the modal with a label.

4. In line 88, the _onClick_ handler for the refresh button should be wrapped in a _try-catch_ block. This is because an error in the _getLeaderboardData_ function can cause the application to crash.

5. In line 93, the _onClick_ handler for the info button should have an _aria-label_ attribute to provide a label for the button. This improves accessibility by providing a label for assistive technology users.

6. In lines 71-77, the _window.open_ method is used to open a new window. This method can be blocked by pop-up blockers in some browsers. A better approach would be to use the _Link_ component from _react-router-dom_ to open the dashboard in the same window.

7. In line 20, the _useDeepEffect_ function is defined. This function is not part of the React API and should be defined outside the component.

8. In line 11, the _lodash_ library is imported. This library is not used in the code and should be removed.

Here is the improved code:


import React, { useEffect, useState, useRef } from 'react';
import './Leaderboard.css';
import { Link } from 'react-router-dom';
import { Table, Progress, Modal, ModalBody, ModalFooter, ModalHeader, Button } from 'reactstrap';

function useDeepEffect(effectFunc, deps) {
  const isFirst = useRef(true);
  const prevDeps = useRef(deps);
  useEffect(() => {
    try {
      const isSame = prevDeps.current.every((obj, index) => isEqual(obj, deps[index]));
      if (isFirst.current || !isSame) {
        effectFunc();
      }
      isFirst.current = false;
      prevDeps.current = deps;
    } catch (error) {
      console.error(error);
    }
  }, deps);
}

const LeaderBoard = ({
  getLeaderboardData,
  getOrgData,
  leaderBoardData,
  loggedInUser,
  organizationData,
  timeEntries,
  asUser,
  setLeaderData,
}) => {
  const userId = asUser || loggedInUser.userId;

  useDeepEffect(() => {
    getLeaderboardData(userId);
    getOrgData();
  }, [timeEntries]);

  useEffect(() => {
    try {
      setLeaderData(leaderBoardData);
    } catch (error) {
      console.error(error);
    }
  }, [leaderBoardData]);

  const [isOpen, setOpen] = useState(false);
  const toggle = () => setOpen(isOpen => !isOpen);

  // add state hook for the popup the personal's dashboard from leaderboard
  const [isDashboardOpen, setIsDashboardOpen] = useState(false);
  const dashboardToggle = item => setIsDashboardOpen(item.personId);
  const dashboardClose = () => setIsDashboardOpen(false);
  const showDashboard = item => {
    dashboardClose();
    window.open(
      `/dashboard/${item.personId}`,
      'Dashboard',
      'toolbar=no, location=no, statusbar=no, menubar=no, scrollbars=

**Comment by GKD890** [comment]
Just add a label in case anyone review first and then find the comment.

**Review by yoyoweiii — APPROVED** [review]
Hi @juneluo, the timelog shows the correct output. Great works! 

Admin account
HGN APP - Personal - Microsoft​ Edge 5_18_2023 6_23_10 PM
CareerInfo xlsx - Google Sheets and 24 more pages - Personal - Microsoft​ Edge 5_18_2023 6_29_03 PM
Volunteer account
HGN APP - Personal - Microsoft​ Edge 5_18_2023 6_29_28 PM
HGN APP - Personal - Microsoft​ Edge 5_18_2023 6_29_36 PM

**Comment by wang9hu** [comment]
Hi @juneluo, I tested your changes and it feels faster and more consistent. I added one console.log to log 're-rendered' for each update, and during one refresh, the times Dashboard re-rendered are way less than before. Great job!
before: 
Screenshot 2023-05-19 001649
after:
Screenshot 2023-05-19 001707

**Comment by papiasharmin** [comment]
Hello @juneluo, the loading time has been minimized. Everything seems to work faster and correctly. I have console log from different pages, initial rendering for timelog component quite a few. great work.

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/86232710/a4378fbd-9fe1-4c13-b0ae-b0d0c3711129

**Comment by EdwinLauM** [comment]
Hi @juneluo. Tested different components. The dashboard re-rendering is definitely less. Timelog component re-renders quite a bit of times.

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/42288987/66678074-37ca-4c80-9aba-78929f8d1e04

**Review by EdwinLauM — APPROVED** [review]
Hi @juneluo. Tested different components. The dashboard re-rendering is definitely less. Timelog component re-renders quite a bit of times.

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/42288987/66678074-37ca-4c80-9aba-78929f8d1e04

**Review by drewlong314 — APPROVED** [review]
I tested the changes and it seems like it is faster.

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/78223925/29c61e2a-f495-41ce-81f4-513724af550b

**Comment by GKD890** [comment]
It does works faster. will come back and approved the code though. Although I'm not sure if I'm adding `console` to the same position as other reviewers. 

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/57503876/9311dcc9-9767-42d1-a09e-97900dd3308f

**Review by eduhorta — APPROVED** [review]
I made a post on Slack talking about the performance and I was having some inconsistent behaviours. Jae mentioned about this PR and now after testing it, seems I have more consistency while rendering the tasks/leaderboard.
I saw other members videos and everything was so fast! Mine is not that much, but, better than before! 👍
Nice job!

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/105796041/c2fe68d9-f052-41f1-92da-b3072b15e3da

**Comment by one-community** [comment]
Thank you everyone, moving this to final review!

**Comment by Aishwaryak01** [comment]
Great and commendable work @juneluo . There is a significant change in the loading time.
I think it would be great if we add a few more comments wherever a function is created or for any important check/step.
Overall code looks good to me.Approved!

**Pull request merged**
